### ISSUE TITLE
Disconnect / Reconnect

### ISSUE THREAD

**Issue by aswanevelder** [aspnet/SignalR#1057]
Hi guys,

Are there javascript events available yet for Disconnect and Reconnect as was available in SignalR 2?

Thanks

**Comment by AustinFelipe** [comment]
@aswanevelder I don't think so, at least I couldn't find it. I'm doing it using ```onclose``` event and just creating a brand new ```HubConnection``` 🤔

**Comment by aswanevelder** [comment]
Thanks Austin, I now resolved just using websockets out of the box, this SignalR project is taking too long, I've been waiting for it since Core 1.0. A pity though they are doing great work, I will maybe re-look at this a few years from now, hope it will be released then. https://docs.microsoft.com/en-us/aspnet/core/fundamentals/websockets

**Comment by moozzyk** [comment]
There is no reconnect in this version of SignalR so there is no event. `onclose` is called when the connection is closed. 
You can now use websockets to connect to SignalR (i.e. no client is needed) - [here](https://github.com/aspnet/SignalR-samples/commit/85c8d9335da66009459434ca468f27d96dc27cb8) is an example (which I need to finish and merge but it should give you an idea)

**Comment by davidfowl** [comment]
We're looking at adding the automatic reconnect functionality back to SignalR in the next preview release.

**Comment by aswanevelder** [comment]
Thanks guys, @davidfowl looking forward to it!

**Comment by KirilRusev** [comment]
@davidfowl sounds great! Do you have road map for the next few months?

**Comment by AwsomeCode** [comment]
@davidfowl @aswanevelder @KirilRusev  
Please don't add reconnect. Its kind of confusing to manage. If client get disconnected, client can initiate reconnect. Its better that server should not manage reconnects.

**Comment by davidfowl** [comment]
It’s all client side. All we’re going to do is make sure the connection is reusable and there’s a way to auto reconnect. The server isn’t managing anything

**Comment by anurse** [comment]
Restarting the connection is now supported in the .NET client (in PR #1147).

#1171 is tracking adding support to the TypeScript client.

Given that, I'm going to close this issue. Feel free to comment if you still have concerns!

**Comment by BrennanConroy** [comment]
Reopening, we are going to be adding automatic reconnect

Closed event called once reconnect tries are done
Need to decide on number of reconnect tries
Add configurability
Back-off delay between reconnect tries
StopAsync stops reconnect
Keep connection re-usability

**Comment by Frank802** [comment]
Hi guys, I'm developing a local dashboard web app that fetch data from many Rpi connected to the same local network. I'm using HubConnection on client side, with a stream subscribe. It works very well except for the fact that sometimes when I navigate to another page and then return to the dashboard page, I got this error: "cannot send data if the connection is not in the 'connected' state". I don't understand why and if I try to refresh the page I got the same error, but if I try to refresh the page using CTRL+F5 (clear cache) It works. Any Ideas?

**Comment by Frank802** [comment]
After some tests, I discover that the problem was a 'timing problem'. I was checking the state of the streaming in the windows on load event where the signalR connection sometime was still in connecting state, that's why sometimes I got the error. I will solve this and then I will post a solution just in case someone else could be face the same problem in the future. Thanks!

**Comment by kmorcinek** [comment]
Disussion is mainly about JavaScript client I guess. 
Are you also planning to implement Reconnect in .NET Client? Maybe I should start new issue for that?

**Comment by anurse** [comment]
When we implement automatic reconnect (just to clarify, manual reconnecting is possible today by simply calling `StartAsync` after the `Closed` event fires), it will be in both clients.

**Comment by radenkozec** [comment]
@aswanevelder Guys can you take a look at my implementation of automatic reconnect and test it to see if it works for you. https://www.radenkozec.com/net-core-signalr-automatic-reconnects/

**Comment by anurse** [comment]
@radenkozec The sample you posted looks good, but you shouldn't have to recreate the HubConnection each time. If you're seeing issues with reusing the same HubConnection instance, let us know in a new issue.

**Comment by radenkozec** [comment]
@anurse Thanks for feedback. I will improve my code and re-use the same HubConnection instance.

**Comment by ruibasecone** [comment]
I was about to ask clarifications about how to handle reconnect on the typescript client but I found this in the meantime: https://github.com/aspnet/Docs/issues/6757

I must say your release process makes me a bit confused but maybe it's just me. Probably it doesn't help (me) that you have .NET and Typescript clients in the same repository. Talking about the Typescript client here, you seem to have some 2.x versions finalised but latest version on npm is 1.0.3, so I don't even know if I will be able to reconnect with 1.03.

**Comment by anurse** [comment]
> you seem to have some 2.x versions finalised

The SignalR versions should be consistent across NuGet and NPM. The issue you're referring to is that SignalR is part of ASP.NET Core. The entire ASP.NET Core platform is at version 2.1 (with version 2.2 being developed now). However, the version of SignalR included in those builds are 1.0 (and 1.1 in the next release).

We recongize this is a bit confusing and are planning to resolve this in the next major version when we can align builds.

A handy reference table:

| ASP.NET Core Version | SignalR NuGet/NPM Package Version |
| - | - |
| 2.1.x | 1.0.x |
| 2.2.x | 1.1.x |

**Comment by ruibasecone** [comment]
@anurse Thank you for the clarification.
